File: app/src/main/java/com/plbear/iweight/model/details/DetailsActivity.java
package com.plbear.iweight.model.details;

import android.app.AlertDialog;
import android.content.Context;
import android.database.ContentObserver;
import android.graphics.Color;
import android.os.Handler;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;

import com.plbear.iweight.R;
import com.plbear.iweight.base.BaseActivity;
import com.plbear.iweight.base.Constant;
import com.plbear.iweight.data.Data;
import com.plbear.iweight.data.DataManager;
import com.plbear.iweight.utils.Utils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Timer;
import java.util.TimerTask;

import butterknife.BindView;

/**
 * Created by yanyongjun on 2018/6/28.
 */

public class DetailsActivity extends BaseActivity {
    @BindView(R.id.btn_details_select_all)
    Button mBtnSelectAll;

    @BindView(R.id.btn_details_change)
    Button mBtnChange;

    @BindView(R.id.btn_details_delete)
    Button mBtnDelete;

    @BindView(R.id.listview_details)
    ListView mListView;

    private DetailsAdapter mAdapter;
    private ArrayList<Data> mDataList;
    private DetailsAdapter.OnItemClick mCallback = new DetailsAdapter.OnItemClick() {
        @Override
        public void itemClick(HashMap<Integer, Boolean> mSelectMap, int selectCount) {
            int size = mDataList.size();
            if (size != selectCount) {
                mBtnSelectAll.setText(R.string.details_title_select_all);
            } else {
                mBtnSelectAll.setText(R.string.details_title_disselect_all);
            }
            if (selectCount == 1) {
                mBtnChange.setClickable(true);
                mBtnChange.setTextColor(Color.BLACK);
            } else {
                mBtnChange.setClickable(false);
                mBtnChange.setTextColor(getResources().getColor(R.color.details_item_bg));
            }
            if (selectCount > 0) {
                mBtnDelete.setClickable(true);
                mBtnDelete.setTextColor(Color.BLACK);
            } else {
                mBtnDelete.setClickable(false);
                mBtnDelete.setTextColor(getResources().getColor(R.color.details_item_bg));
            }
        }

        @Override
        public void longItemClick(boolean editMode) {
            if (editMode) {
                mBtnSelectAll.setVisibility(View.VISIBLE);
                mBtnChange.setVisibility(View.VISIBLE);
                mBtnDelete.setVisibility(View.VISIBLE);
                mBtnSelectAll.setText(R.string.details_title_select_all);
                mBtnChange.setClickable(false);
                mBtnChange.setTextColor(getResources().getColor(R.color.details_item_bg));
                mBtnDelete.setClickable(false);
                mBtnDelete.setTextColor(getResources().getColor(R.color.details_item_bg));
            } else {
                mBtnChange.setVisibility(View.GONE);
                mBtnDelete.setVisibility(View.GONE);
            }
        }
    };

    private ContentObserver mObserver = new ContentObserver(new Handler()) {
        @Override
        public void onChange(boolean selfChange) {
            mAdapter.notifyDataSetChanged();
            super.onChange(selfChange);
        }
    };

    @Override
    public int getLayout() {
        return R.layout.activity_details;
    }

    @Override
    public void afterLayout() {
        init();
    }

    @Override
    protected void onStop() {
        getContentResolver().unregisterContentObserver(mObserver);
        super.onStop();
    }

    @Override
    protected void onStart() {
        getContentResolver().registerContentObserver(Constant.CONTENT_URI, true, mObserver);
        super.onStart();
    }

    private void init() {
        mDataList = DataManager.getInstance().queryAll();
        Utils.sortDataBigToSmall(mDataList);
        mAdapter = new DetailsAdapter(this, mDataList, mCallback);
        mListView.setAdapter(mAdapter);
        mListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {

            }
        });

        mBtnSelectAll.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mAdapter.isEditMode) {
                    mAdapter.onSelectAllClick();
                } else {
                    mAdapter.enterEditMode();
                }
            }
        });

        mBtnSelectAll.setText("编辑");
        mBtnChange.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                showChangeDialog();
            }
        });
        mBtnDelete.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                showDeleteDialog();
            }
        });
    }

    private void showDeleteDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        View layout = getLayoutInflater().inflate(R.layout.dialog_details_delete, null);
        builder.setView(layout);
        final AlertDialog dialog = builder.create();
        Button btnSubmit = (Button) layout.findViewById(R.id.dialog_submit);
        btnSubmit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ArrayList<Data> list = mAdapter.getSelectData();
                DataManager.getInstance().delete(list);
                dialog.dismiss();
                mAdapter.exitEditMode();
            }
        });

        Button btnCancal = layout.findViewById(R.id.dialog_cacnel);
        btnCancal.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dialog.dismiss();
            }
        });

        dialog.show();
    }

    private void showChangeDialog() {
        ArrayList<Data> list = mAdapter.getSelectData();
        final Data data = list.get(0);
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        final View layout = getLayoutInflater().inflate(R.layout.dialog_main_input_weight, null);
        builder.setView(layout);

        final AlertDialog dialog = builder.create();
        Button btnSubmit = (Button) layout.findViewById(R.id.dialog_submit);
        final EditText editText = layout.findViewById(R.id.dialog_input_weight);
        editText.setText("" + data.getWeight());
        editText.setSelection(editText.getText().length());

        btnSubmit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                try {
                    float weight = Float.parseFloat(editText.getText().toString());
                    if (weight < 5 || weight > 200) {
                        Utils.showToast("您输入的值太不合理了");
                        return;
                    }
                    data.setWeight(weight);
                    DataManager.getInstance().update(data);
                    dialog.dismiss();
                    mAdapter.exitEditMode();
                } catch (Exception e) {
                    e.printStackTrace();
                    Utils.showToast("请检查输入的值");
                }
            }
        });

        Button btnCancel = layout.findViewById(R.id.dialog_cacnel);
        btnCancel.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dialog.dismiss();
            }
        });

        dialog.show();

        editText.requestFocus();
        editText.setFocusable(true);
        editText.setFocusableInTouchMode(true);
        Timer timer = new Timer();
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                if (editText != null) {
                    InputMethodManager input = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
                    input.toggleSoftInput(0, InputMethodManager.SHOW_FORCED);
                }
            }
        }, 200);


    }

     public void onBackPressed() {
         DetailsAdapter adapter = (DetailsAdapter) mListView.getAdapter();
        if(adapter.isEditMode){
            adapter.exitEditMode();
            mBtnSelectAll.setText("编辑");
            return;
        }
        super.onBackPressed();
    }

}

File: app/src/main/java/com/plbear/iweight/model/details/DetailsActivity.kt
package com.plbear.iweight.model.details

import android.app.AlertDialog
import android.content.Context
import android.database.ContentObserver
import android.graphics.Color
import android.net.Uri
import android.os.Handler
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.AdapterView
import android.widget.Button
import android.widget.EditText
import android.widget.Toast

import com.plbear.iweight.R
import com.plbear.iweight.base.Constant
import com.plbear.iweight.data.DataManager
import com.plbear.iweight.base.BaseActivity
import com.plbear.iweight.data.Data
import com.plbear.iweight.utils.MyLog
import com.plbear.iweight.utils.Utils
import kotlinx.android.synthetic.main.activity_details.*
import kotlinx.android.synthetic.main.include_title.*

import java.util.ArrayList
import java.util.HashMap
import java.util.Timer
import java.util.TimerTask

/**
 * Created by yanyongjun on 2016/11/23.
 */

class DetailsActivity : BaseActivity() {

    lateinit var mAdapter: DetailsAdapter
    private var mDataList: ArrayList<Data>? = null
    private val mCallback = object : DetailsAdapter.OnItemClick {
        override fun itemClick(mSelectMap: HashMap<Int, Boolean>, selectCount: Int) {
            val size = mDataList!!.size
            if (size != selectCount) {
                btn_details_select_all.setText(R.string.select_all)
            } else {
                btn_details_select_all.setText(R.string.disselect_all)
            }
            if (selectCount == 1) {
                btn_details_change.isClickable = true
                btn_details_change.setTextColor(Color.BLACK)
            } else {
                btn_details_change.isClickable = false
                btn_details_change.setTextColor(resources.getColor(R.color.details_item_bg))
            }
            if (selectCount > 0) {
                btn_details_delete.isClickable = true
                btn_details_delete.setTextColor(Color.BLACK)
            } else {
                btn_details_delete.isClickable = false
                btn_details_delete.setTextColor(resources.getColor(R.color.details_item_bg))
            }
        }

        override fun longItemClick(editMode: Boolean) {
            if (editMode) {
                btn_details_select_all.visibility = View.VISIBLE
                btn_details_change.visibility = View.VISIBLE
                btn_details_delete.visibility = View.VISIBLE
                btn_details_select_all.setText(R.string.select_all)
                btn_details_change.isClickable = false
                btn_details_change.setTextColor(resources.getColor(R.color.details_item_bg))
                btn_details_delete.isClickable = false
                btn_details_delete.setTextColor(resources.getColor(R.color.details_item_bg))
            } else {
                btn_details_change.visibility = View.GONE
                btn_details_delete.visibility = View.GONE
            }
        }
    }

    private val mObserver = object : ContentObserver(Handler()) {
        override fun onChange(selfChange: Boolean, uri: Uri) {
            mAdapter!!.notifyDataSetChanged()
            super.onChange(selfChange, uri)
        }
    }

    override fun getLayout(): Int {
        return R.layout.activity_details
    }

    override fun afterLayout() {
        init()
    }

    override fun onStop() {
        this.contentResolver.unregisterContentObserver(mObserver)
        super.onStop()
    }

    override fun onStart() {
        MyLog.e(TAG, "onStart1")
        this.contentResolver.registerContentObserver(Constant.CONTENT_URI, true, mObserver)
        MyLog.e(TAG, "onStart normal_2")
        super.onStart()
    }

    private fun init() {
        mDataList = DataManager.getInstance().queryAll()
        Utils.sortDataBigToSmall(mDataList!!)
        mAdapter = DetailsAdapter(this, mDataList!!, mCallback)
        listview_details.adapter = mAdapter
        listview_details.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id -> }

        //初始化标题栏
        btn_details_select_all.setOnClickListener {
            if (mAdapter.isEditMode) {
                mAdapter.onSelectAllClick()
            } else {
                mAdapter.enterEditMode()
            }
        }
        btn_details_select_all.setText("编辑")

        btn_details_change.setOnClickListener { showChangeDialog() }
        btn_details_delete.setOnClickListener { showDeleteDialog() }
    }

    fun onClick_back(v: View) {
        finish()
    }

    private fun showDeleteDialog() {
        val builder = AlertDialog.Builder(this)
        val inflater = layoutInflater
        val layout = inflater.inflate(R.layout.dialog_details_delete, null)
        builder.setView(layout)
        val dialog = builder.create()
        val btnSubmit = layout.findViewById<View>(R.id.dialog_submit) as Button
        btnSubmit.setOnClickListener {
            val list = mAdapter!!.selectData
            DataManager.getInstance().delete(list)
            dialog.dismiss()
            mAdapter!!.exitEditMode()
        }

        val btnCancel = layout.findViewById<Button>(R.id.dialog_cacnel)
        btnCancel.setOnClickListener { dialog.dismiss() }
        dialog.show()
    }

    private fun showChangeDialog() {
        val list = mAdapter.selectData
        MyLog.e(TAG, "yanlog list $list")
        val data = list[0]
        val builder = AlertDialog.Builder(this)
        val inflater = layoutInflater
        val layout = inflater.inflate(R.layout.dialog_main_input_weight, null)
        builder.setView(layout)
        val dialog = builder.create()
        val btnSubmit = layout.findViewById<Button>(R.id.dialog_submit)
        val editText = layout.findViewById<EditText>(R.id.dialog_input_weight)
        editText.setText("" + data.weight)
        editText.setSelection(editText.text.length)
        btnSubmit.setOnClickListener(View.OnClickListener {
            try {
                val weight = java.lang.Float.parseFloat(editText.text.toString())
                if (weight < 5 || weight > 200) {
                    Toast.makeText(this@DetailsActivity, "您输入的值太不合理了，在逗我玩吧~", Toast.LENGTH_SHORT).show()
                    return@OnClickListener
                }
                data.weight = weight
                DataManager.getInstance().update(data)
                dialog.dismiss()
                mAdapter!!.exitEditMode()
            } catch (e: Exception) {
                e.printStackTrace()
                Toast.makeText(this@DetailsActivity, "输入值不合法，请重新输入~", Toast.LENGTH_SHORT).show()
            }
        })

        val btnCancel = layout.findViewById<Button>(R.id.dialog_cacnel)
        btnCancel.setOnClickListener { dialog.dismiss() }
        dialog.show()

        editText.requestFocus()
        editText.isFocusable = true
        editText.isFocusableInTouchMode = true
        val timer = Timer()
        timer.schedule(object : TimerTask() {
            override fun run() {
                if (editText != null) {
                    val imm = this@DetailsActivity.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                    imm.toggleSoftInput(0, InputMethodManager.SHOW_FORCED)
                }
            }
        }, 200)
    }

    override fun onBackPressed() {
        val adapter = listview_details.adapter as DetailsAdapter
        if (adapter.isEditMode) {
            adapter.exitEditMode()
            btn_details_select_all.setText("编辑")
            return
        }
        super.onBackPressed()
    }


}
